content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+expression like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }`
+to enforce a certain Go compiler version.
+
+## 2. Benefit: overrides
+
+As a consequence from the parametrized builds, we can also change the value of
+the parameters after the fact, using the derivations `override` function.
+
+Consider this new `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  hello = pkgs.callPackage ./hello.nix { };
+  people = pkgs.callPackage ./hello.nix { audience = "people"; };
+  folks = hello.override { audience = "folks"; };
+}
+```
+
+Building and running the `folks` attribute, will give again a new version of the
+script.
+
+All the other parameters will remain the same as the have been when `hello` was
+instantiated.
+
+This is especially useful and often seen on packages that provide a whole lot of
+options to optimize the build.
+
+An example to mention here is the `neovim` attribute in nixpkgs, which has has 
+some overrideable arguments like `extraLuaPackages`, `extraPythonPackages`, or
+`withRuby`.
+
+## 3. Benefit: modifiable
+
+And now I want to introduce one of my most favorite benefits:
+
+You can actually create your own version of `callPackage`, which comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+```nix
+# In the next examples I will not implement or show the "called" 
+# files, as I think they are not necessary to understand the point I
+# want to make.
+```
+
+Consider the following initial version:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  a = pkgs.callPackage ./a.nix { };
+  b = pkgs.callPackage ./b.nix { inherit a; };
+  c = pkgs.callPackage ./c.nix { inherit b; };
+  d = pkgs.callPackage ./d.nix { };
+  e = pkgs.callPackage ./e.nix { inherit c d; };
+}
+```
+
+Here you have to remember passing required arguments that are not in nixpkgs'
+toplevel manually.
+
+This can become quite tedious quickly, especially the larger the set becomes.
+
+Therefore we can use `lib.callPackageWith` to create our own `callPackage` version.
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+  callPackage = lib.callPackageWith (pkgs // packages);
+  packages = {
+    a = callPackage ./a.nix { };
+    b = callPackage ./b.nix { };
+    c = callPackage ./c.nix { };
+    d = callPackage ./d.nix { };
+    e = callPackage ./e.nix { };
+  };
+in
+  packages
+```
+
+Our modified `callPackage` now will exactly "know" how to resolve the dependencies
+through the set defined by `pkgs // packages`.
+
+Nix' laziness does us a good favour here and makes this actually possible.
+
+## Summary
+
+So using `callPackage` doesn't only make your code uniform to what you see a lot
+in nixpkgs already, it also gives you some things for free:
+
+1. parametrized builds
+2. overrideable builds
+3. cleaner implementation of large interdepending package sets
```

Comment:
```suggestion
Using `callPackage` does not only follow `nixpkgs` conventions, making your code easy to follow for experienced Nix users. It also gives you some things for free:

1. parametrized builds
2. overridable builds
3. robust implementation of large interdependent package sets
```

I don't know if the recursive method is all that much cleaner (which is subjective anyway in this case - having less to type but making it significantly harder to evaluate in your head does not follow my intuition of clean code). I tried to convey that in the suggestion, selling it as a trade-off, so that trade-off's pro side should be reflected here.